app/electron/main.ts

```diff
@@ -383,6 +402,63 @@ function startElecron() {
     });
 
     if (!useExternalServer) {
+      const runningHeadlamp = await getRunningHeadlampPIDs();
+      if (!!runningHeadlamp) {
+        const resp = dialog.showMessageBoxSync(mainWindow, {
+          // Avoiding mentioning Headlamp here because it may run under a different name depending on branding (plugins).
+          title: i18n.t('Other process running'),
+          message: i18n.t(
+            'Looks like another process is already running. Continue by stopping that process automatically, or quit?'
+          ),
+          type: 'question',
+          buttons: [i18n.t('Continue'), i18n.t('Quit')],
+        });
+
+        if (resp === 0) {
+          runningHeadlamp.forEach(killProcess);
+        } else {
+          mainWindow.close();
+          return;
+        }
+      }
+
+      // If we reach here, then we attempted to kill headlamp-server. Let's make sure it's killed
+      // before starting our own, or else we may end up in a race condition (failing to start the
+      // new one before the existing one is fully killed).
+      if (!!runningHeadlamp) {
+        let stillRunning = true;
+        const waitTime = 250; //ms
+        const attempts = (10 * 1000) / 250; // 4 times per sec, 10 secs max
+        for (let tries = 0; tries < attempts && stillRunning; tries++) {
+          console.debug(
+            `Checking if Headlamp is still running after we asked it to be killed; try=${
+              tries + 1
+            }/${attempts}`
+          );
+          // Wait 250ms
+          await setTimeout(() => {}, waitTime);
```

Comment:
For the most part, I guess the process will be killed in much less than 250ms. An improvement could be to not wait before checking the first time. Another improvement could be to back off a bit more each time (2ms, 4ms, 8ms, 16ms, 32ms, 64ms, 128ms) so as to reduce the wait time.

All that said, 250ms is not so long to wait... so this seems fine to me.

Reply:
There was a race between killing + the process effectively getting killed (maybe my computer is too fast). I can do an iterative back-off but I don't think it's worth anything <10ms...